soc: arm: samv71: Initial support

nandojve opened the pull request:
The Atmel SAMV71 is a Cortex M7 that shares it's peripherals with the sam family. It's the full featured version of this series. This PR add support for the SAMV71 MCU and the SAMV71 Xplained Ultra development board.

![Atmel_Cortex_M7_MCU](https://user-images.githubusercontent.com/7027887/73592493-e2176f00-44d9-11ea-8372-09db9aacff09.png)

Since most of drivers are ready there are many things that can be used just after add basic support (DTS, SOC, Board). Some drivers need cosmetic changes to be enabled and PR are provided and others need PR to do similar adjusts - future work.

This PR focus on add the basics - DTS, SOC and Board.

What works:

- [ ] UART (need PR - Future)
- [X] USART
- [X]  SPI #22394, #22395
- [X]  I2C
- [ ] I2S (need PR - Future)
- [X]  GPIO #22396
- [X]  XDMAC
- [ ]  RTC
- [X]  Watchdog
- [x]  ADC #22398
- [x]  USB #22397
- [X] Entropy
- [X]  Ethernet GMAC
- [ ]  Timers
- [X] Flash
- [X] PWM

requires zephyrproject-rtos/hal_atmel/pull/3

nandojve committed:
dts: arm: atmel: Add samv71 device tree

Add device tree for Atmel samv71 SoC.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
soc: arm: atmel: Add samv71 mcu

Add samv71 related soc files.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
boards: arm: Add sam_v71_xult board

Add sam_v71_xult board files. This enable use of Atmel samv71 SoC.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
CODEOWNERS: add owner for the samv71 platform

Add recent collaborator as codeowner.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve commented:
Hi folks,

It is failing shippable because of dependency zephyrproject-rtos/hal_atmel/pull/3 that need be accepted.

mnkp commented:
This PR can (should in fact) modify `west.yml` to pull in correct version of the Atmel hal library.

stephanosio commented:
> This PR can (should in fact) modify `west.yml` to pull in correct version of the Atmel hal library.

https://docs.zephyrproject.org/latest/guides/modules.html#changes-to-existing-modules

stephanosio commented on west.yml:
```diff
@@ -26,7 +26,7 @@ manifest:
   projects:
     - name: hal_atmel
-      revision: 5690f5b84495c8b657ff204c5c827df1ab9e12e3
+      revision: 72443ed8f1fa39dc3c1c2a5dba0958381106036d
```
Need to do `revision: pull/3/head` here because the PR branch is not fetched into the upstream repo.

mnkp commented on boards/arm/sam_v71_xult/sam_v71_xult.dts:
```diff
@@ -0,0 +1,171 @@
+		compatible = "gpio-leds";
+		yellow_led0: led_0 {
+			gpios = <&porta 23 0>;
+			label = "User LED";
```
`"User LED 0"`

mnkp commented on boards/arm/sam_v71_xult/sam_v71_xult.dts:
```diff
@@ -0,0 +1,171 @@
+		};
+		yellow_led1: led_1 {
+			gpios = <&portc 9 0>;
+			label = "User LED";
```
`"User LED 1"`

mnkp commented on boards/arm/sam_v71_xult/sam_v71_xult.dts:
```diff
@@ -0,0 +1,171 @@
+	 * partition, the code will be linked
+	 * to, and restricted to that
+	 * partition.
+	 */
```
This has been copied from the current tree, however the description is not valid anymore #13151. I propose to simply remove it.

mnkp commented on boards/arm/sam_v71_xult/sam_v71_xult_defconfig:
```diff
@@ -0,0 +1,18 @@
+CONFIG_USART_SAM=y
+CONFIG_USART_SAM_PORT_1=y
+CONFIG_BOARD_SAM_V71_XPLAINED=y
+CONFIG_BUILD_OUTPUT_HEX=y
```
This has been added before as a workaround. In a meantime the build system has been modified to always build hex. I propose to remove it. As a long term solution we should fix openocd scripts to reliably flash bin files.

mnkp commented on boards/arm/sam_v71_xult/sam_v71_xult.dts:
```diff
@@ -0,0 +1,171 @@
+		 * on docs
+		 */
+		sw0_user_button: button_1 {
+			label = "User Button";
```
`"User Button 0"`

mnkp commented on boards/arm/sam_v71_xult/sam_v71_xult.dts:
```diff
@@ -0,0 +1,171 @@
+					    GPIO_INT_ACTIVE_LOW)>;
+		};
+		sw1_user_button: button_2 {
+			label = "User Button";
```
`"User Button 1"`

mnkp reviewed (commented):
I haven't tested it locally but it looks very well overall. Thanks!

I may be missing something but I believe none of the revision B parts is going to compile at the moment. You would need to update `zephyrproject/modules/hal/atmel/asf/sam/include/CMakeLists.txt` in zephyrproject-rtos/hal_atmel/pull/3.

nandojve commented:
> I haven't tested it locally but it looks very good overall. Thanks!
> 
> I may be missing something but I believe none of the revision B parts is going to compile at the moment. You would need to update `zephyrproject/modules/hal/atmel/asf/sam/include/CMakeLists.txt` in [zephyrproject-rtos/hal_atmel/pull/3](https://github.com/zephyrproject-rtos/hal_atmel/pull/3).

I added solution for revision B at CMakeLists.txt. When I tryied build and I've detected errors with deprecated include files. I removed that files and includes. After that, I detected that samv71b/component/usart.h differ from others usart definitions and cause usart driver do not build. We need make a decision about add  missing symbols or patch usart driver. In general, I've been add missing symbols because we don't need touch on driver to maintain clean code.

mnkp commented:
> I detected that samv71b/component/usart.h differ from others usart definitions and cause usart driver do not build. We need make a decision about add missing symbols or patch usart driver. In general, I've been add missing symbols because we don't need touch on driver to maintain clean code.

Unfortunately, the quality of the header files provided by Atmel has never been very good. I opened a few tickets with Atmel in the past (when they were still Atmel). There were engineers there trying to do a good job and fix things but the new versions of packs that got released still contained bugs and were often inconsistent. I see that under Microchip that didn't change.

I had a look at the errors. They are caused by inconsistent naming of bit-fields. In this case yes, I believe that updating the header files is the best course of action. It would also be good to open a ticket with Microchip. They should do better.

Just a note: any custom modifications to original header files should be done in a separate commit. It makes it easier to see what was changed and helps maintaining updates.

nandojve commented:
> > I detected that samv71b/component/usart.h differ from others usart definitions and cause usart driver do not build. We need make a decision about add missing symbols or patch usart driver. In general, I've been add missing symbols because we don't need touch on driver to maintain clean code.
> 
> Unfortunately, the quality of the header files provided by Atmel has never been very good. I opened a few tickets with Atmel in the past (when they were still Atmel). There were engineers there trying to do a good job and fix things but the new versions of packs that got released still contained bugs and were often inconsistent. I see that under Microchip that didn't change.
> 
> I had a look at the errors. They are caused by inconsistent naming of bit-fields. In this case yes, I believe that updating the header files is the best course of action. It would also be good to open a ticket with Microchip. They should do better.
> 
> Just a note: any custom modifications to original header files should be done in a separate commit. It makes it easier to see what was changed and helps maintaining updates.

Done!

dbkinder commented on boards/arm/sam_v71_xult/doc/index.rst:
```diff
@@ -0,0 +1,181 @@
+Serial Port
+===========
+
+The ATSAMV71Q21 MCU has five UARTs and three USARTs. One of the USARTs is
```
Document which one is configured for the console here:  ``USART1 is configured ...``

dbkinder commented on boards/arm/sam_v71_xult/doc/index.rst:
```diff
@@ -0,0 +1,181 @@
+  Gateway Interface (DGI)
+- JTAG interface connector
+- One reset and two user pushbutton
+- Two yellow user LED
```
change to ``pushbuttons`` and ``LEDs`` (since there are two of them)

dbkinder commented on boards/arm/sam_v71_xult/doc/index.rst:
```diff
@@ -0,0 +1,181 @@
+Support for Atmel SAM E microcontroller series was added in OpenOCD release
+0.10.0, which was added in Zephyr SDK 0.9.2.
+
+By default a factory new SAM V71 chip will boot SAM-BA boot loader located in
```
Let's add a link to the document since you have defined it at the bottom of the page:
```
 ... will boot the `SAM-BA`_ boot loader ...
```

nandojve replied:
Thank you for all your support :D

dbkinder reviewed (changes requested):
docs look ok with a few tweaks...

nandojve commented:
Rebase and apply #21137 and #21134, from my side I believe is everything Ok.

mnkp reviewed (commented):
#21137 got closed, we should drop "soc: samv71: Disable watchdog when watchdog driver is disabled" commit. Otherwise all looks good. Thanks!

nandojve commented:
Rabase and drop **soc: samv71: Disable watchdog when watchdog driver is disabled**.

dbkinder reviewed (approved):
docs LGTM, thanks!

nandojve commented:
Rebase and apply:

dts: Rename generated_dts_board*.{h,conf} to devicetree*.{h,conf}
boards: sam: kconfig: Do not assign promptless SOC_FAMILY_SAM symbol**

nandojve commented:
Rebase and apply:

soc: arm: atmel_sam: Select CPU DWT feature symbol

nashif commented:
DNM until west.yml is updated

nandojve committed:
west.yml: update hal_atmel for SAM4E/SAMV71

Update hal_atmel to include vendor files for SAM4E and SAMV71 series.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>
